### ISSUE TITLE
Use TypeScript?

### ISSUE THREAD

**Issue by glebm** [katspaugh/wavesurfer.js#1209]
Since WaveSurfer already has thorough type annotations, perhaps it would make sense to use TypeScript?

The ES6 and ES5 versions can be generated by the TypeScript compiler and checked in.

There is a tool to convert existing annotated JavaScript to TypeScript here called Gents:
https://github.com/angular/clutz#gents---closure-to-typescript-converter

Documentation generator here:
https://github.com/TypeStrong/typedoc

**Comment by katspaugh** [comment]
The TypeScript compiler works perfectly with regular JavaScript and the type inference, especially with JSDoc annotations. What we might add is a good jsconfig.js similar to [this](https://github.com/katspaugh/language-roulette/blob/master/jsconfig.json).

**Comment by glebm** [comment]
JSDoc type system is very limited, some things just can't be expressed in it.

For example, event handler arguments that differ depending on the argument:

```typescript
export declare class WaveSurfer {
  on(event: 'region-created', listener: (region: WaveSurfer.Region) => {}): void;
  on(event: 'region-click', listener: (region: WaveSurfer.Region, event: MouseEvent) => {}): void;
  on(event: 'region-dblclick', listener: (region: WaveSurfer.Region, event: MouseEvent) => {}): void;
  on(event: 'contextmenu', listener: (event: MouseEvent) => {}): void;
  ...
}
```

The above is from my type declarations for WaveSurfer v1.

**Comment by katspaugh** [comment]
Would it be possible to generate inferred typings, add missing stuff manually and publish in the definitely-typed  repository?

**Comment by glebm** [comment]
It would be possible, but then you'd have to keep them in sync. Why not just write it in TypeScript to begin with? You have type annotations everywhere already anyway, so it's not like it's more effort.

**Comment by mspae** [comment]
I'm all for it, but I think even with automatic conversion it will still require quite a large amount of additional tweaking & annotating. That said, if someone is willing to do it I think it would be awesome.

**Comment by mspae** [comment]
@katspaugh Just to clarify, because it wasn't clear to me from your post: Did you mean it isn't strictly necessary (for an IDE-like experience in an editor) or that you wouldn't want a PR to migrate to typescript?

I'm asking because I'm considering migrating the code and opening a PR and wouldn't start if you don't want it.

I agree that IDE-like experience has gotten much better with better jsdoc comments but I still think migrating to typescript proper would make sense because:
- It enforces people to properly type the code in their PRs (instead of adding JSdoc comments, which are not part of the language itself) and allows our CI build process to statically check types
- As @glebm mentioned, typescript does allow more advanced type annotations than jsdoc style comments.

**Comment by katspaugh** [comment]
@mspae if you want to open a PR – I'm sure the result should be good.

An alternative would be to use something like https://github.com/Microsoft/dts-gen to generate a separate typings file and maintain it in parallel with the ES6 codebase. But I'm not against TypeScript proper. 👍

**Comment by grosch** [comment]
Any progress on this?

**Comment by kubk** [comment]
It looks like this issue can be closed, because typings can be installed using `npm install wavesurfer.js`
